www/waterfall_view/src/module/main.module.js

```diff
@@ -635,14 +640,89 @@ var WaterfallController = (function() {
             });
         }
 
-        renderNewData() {
+        toggleTag(tag) {
+            if (!this.$scope.tags_filter.includes(tag)) {
+                this.$scope.tags_filter.push(tag);
+            } else {
+                this.$scope.tags_filter = _.remove(this.$scope.tags_filter, function(currentTag) {
+                    return currentTag != tag;
+                });
+            }
+            this.$location.search("tags", this.$scope.tags_filter);
+            this.renderNewData(this.$scope.tags_filter);
+        }
+
+        isTagToggled(tag) {
+            if (this.$scope.tags_filter.includes(tag)) {
+                return 'label-success';
+            } else {
+                return 'label-default';
+            }
+        }
+
+        clearTags() {
+            this.$scope.tags_filter = this.tags_filter = [];
+            this.$location.search("tags", this.$scope.tags_filter);
+            this.renderNewData(this.$scope.tags_filter);
+        }
+
+        renderNewData(currentTags) {
+            currentTags = this.$location.search()['tags'];
+            if (currentTags != null) {
+                if (typeof currentTags == 'string') {
+                    this.$scope.tags_filter = this.tags_filter = [currentTags];
+                } else {
+                    this.$scope.tags_filter = this.tags_filter = currentTags;
+                }
+            }
             this.groups = this.dataProcessorService.getGroups(this.all_builders, this.builds, this.c.threshold);
             if (this.s.show_builders_without_builds.value) {
                 this.$scope.builders = this.all_builders;
             } else {
                 this.$scope.builders = (this.builders = this.dataProcessorService.filterBuilders(this.all_builders));
             }
+            var all_tags = [];
+            for (let builder of Array.from(this.builders)) {
```

Comment:
Array.from should rarely be needed. It is often a remain from decaffeinate that never been cleaned up